### PR TITLE
jit: support for various float operations

### PR DESCRIPTION
**Pull request by mathetake**

### CODE REVIEW

File: wasm/wazeroir/interpreter.go [hunk 1]
```diff
@@ -1222,7 +1223,7 @@ func (it *interpreter) callNativeFunc(f *interpreterFunction) {
 						um := math.Abs(float64(f - u))
 						dm := math.Abs(float64(f - d))
 						h := u / 2.0
-						if um < dm || float32(math.Floor(float64(h))) == h {
+						if um < dm || (um == dm && float32(math.Floor(float64(h))) == h) {
```
Comment (mathetake) [inline wasm/wazeroir/interpreter.go:1226]: found a bug in wazeroir interpreter which hasn't been caught by spectest!

File: wasm/jit/jit_amd64.go [hunk 2]
```diff
@@ -23,6 +24,21 @@ import (
 	"github.com/tetratelabs/wazero/wasm/wazeroir"
 )
 
+var (
+	float32SignBitMask        int64 = 1 << 31
+	float32SignBitMaskAddress uintptr
+	// This equals 1 << 63 but Go compiler errors as the number doesn't fit into
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:30]:
```suggestion
	// float64SignBitMask equals 1 << 63 but Go compiler errors as the number doesn't fit into
```

File: wasm/jit/jit_amd64.go [hunk 3]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
 	return nil
 }
 
+// compileAbs emits the instructions to take absolute value of the top value of float type on the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1459]:
```suggestion
// compileAbs adds instructions to replace the top value of float type on the stack with its absolute value.
```

File: wasm/jit/jit_amd64.go [hunk 4]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+	}
+	shiftRightByOne.To.Type = obj.TYPE_REG
+	shiftRightByOne.To.Reg = target.register
+	shiftRightByOne.From.Type = obj.TYPE_CONST
+	shiftRightByOne.From.Offset = 1
+	c.addInstruction(shiftRightByOne)
+	return nil
+}
+
+// compileNeg emits the instructions to negate the top value of float type on the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1498]:
```suggestion
// compileNeg adds instructions to replace the top value of float type on the stack with its negated value.
```

File: wasm/jit/jit_amd64.go [hunk 5]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+	}
+	negate.From.Type = obj.TYPE_REG
+	negate.From.Reg = tmpReg
+	negate.To.Type = obj.TYPE_REG
+	negate.To.Reg = target.register
+	c.addInstruction(negate)
+	return nil
+}
+
+// compileCeil emits the instructions to take the ceil of the top value of float type on the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1543]:
```suggestion
// compileCeil adds instructions to replace the top value of float type on the stack with its ceiling value.
```

File: wasm/jit/jit_amd64.go [hunk 6]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+// For example, stack [..., 1.123] results in [..., 2.0]. This is equivalent to "math.Ceil".
+func (c *amd64Compiler) compileCeil(o *wazeroir.OperationCeil) (err error) {
+	// Internally, ceil can be performed via ROUND instruction with 0x02 mode.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1546]: nice comment+dispatch pattern as it helps me understand.

File: wasm/jit/jit_amd64.go [hunk 7]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+	// See https://android.googlesource.com/platform/bionic/+/882b8af/libm/x86_64/ceilf.S for example.
+	return c.emitRoundInstruction(o.Type == wazeroir.Float32, 0x02)
+}
+
+// compileFloor emits the instructions to take the floor of the top value of float type on the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1551]:
```suggestion
// compileFloor adds instructions to replace the top value of float type on the stack with its floor value.
```

File: wasm/jit/jit_amd64.go [hunk 8]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+	negate.From.Type = obj.TYPE_REG
+	negate.From.Reg = tmpReg
+	negate.To.Type = obj.TYPE_REG
+	negate.To.Reg = target.register
+	c.addInstruction(negate)
+	return nil
+}
+
+// compileCeil emits the instructions to take the ceil of the top value of float type on the stack.
+// For example, stack [..., 1.123] results in [..., 2.0]. This is equivalent to "math.Ceil".
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1544]: I assume you are quoting "math.Ceil" to prevent importing it.

File: wasm/jit/jit_amd64.go [hunk 9]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+
+// compileFloor emits the instructions to take the floor of the top value of float type on the stack.
+// For example, stack [..., 1.123] results in [..., 1.0]. This is equivalent to "math.Floor".
+func (c *amd64Compiler) compileFloor(o *wazeroir.OperationFloor) (err error) {
+	// Internally, floor can be performed via ROUND instruction with 0x01 mode.
+	// See https://android.googlesource.com/platform/bionic/+/882b8af/libm/x86_64/floorf.S for example.
+	return c.emitRoundInstruction(o.Type == wazeroir.Float32, 0x01)
+}
+
+// compileTrunc emits the instructions to take the integer value of the top value of float type on the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1559]:
```suggestion
// compileTrunc adds instructions to replace the top value of float type on the stack with its truncated value.
```

File: wasm/jit/jit_amd64.go [hunk 10]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+// For example, stack [..., 1.9] results in [..., 1.0]. This is equivalent to "math.Trunc".
+func (c *amd64Compiler) compileTrunc(o *wazeroir.OperationTrunc) error {
+	// Internally, trunc can be performed via ROUND instruction with 0x03 mode.
+	// See https://android.googlesource.com/platform/bionic/+/882b8af/libm/x86_64/truncf.S for example.
+	return c.emitRoundInstruction(o.Type == wazeroir.Float32, 0x03)
+}
+
+// compileNearest emits the instructions to take the nearest integer value of the top value of float type on the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1567]:
```suggestion
// compileNearest adds instructions to replace the top value of float type on the stack with its nearest integer value.
```

File: wasm/jit/jit_amd64.go [hunk 11]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+// For example, stack [..., 1.9] results in [..., 2.0]. This is NOT equivalent to "math.Round" and instead has the same
+// the sematics of LLVM's rint instrinsic. See https://llvm.org/docs/LangRef.html#llvm-rint-intrinsic.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1569]:
```suggestion
// the semantics of LLVM's rint instrinsic. See https://llvm.org/docs/LangRef.html#llvm-rint-intrinsic.
```

File: wasm/jit/jit_amd64.go [hunk 12]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+// compileTrunc emits the instructions to take the integer value of the top value of float type on the stack.
+// For example, stack [..., 1.9] results in [..., 1.0]. This is equivalent to "math.Trunc".
+func (c *amd64Compiler) compileTrunc(o *wazeroir.OperationTrunc) error {
+	// Internally, trunc can be performed via ROUND instruction with 0x03 mode.
+	// See https://android.googlesource.com/platform/bionic/+/882b8af/libm/x86_64/truncf.S for example.
+	return c.emitRoundInstruction(o.Type == wazeroir.Float32, 0x03)
+}
+
+// compileNearest emits the instructions to take the nearest integer value of the top value of float type on the stack.
+// For example, stack [..., 1.9] results in [..., 2.0]. This is NOT equivalent to "math.Round" and instead has the same
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1568]: If not equiv to math.round, is there a better example we can find which doesn't result in the same value?

File: wasm/jit/jit_amd64.go [hunk 13]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+	//
+	// 000000000000000f <_wasm_function_1>:
+	// 	f:        55                      push   %rbp
+	//  10:       48 89 e5                mov    %rsp,%rbp
+	//  13:       66 0f 3a 0b c0 00       roundsd $0x0,%xmm0,%xmm0
+	//  19:       48 89 ec                mov    %rbp,%rsp
+	//  1c:       5d                      pop    %rbp
+	//  1d:       c3                      retq
+	//
+	// Here, actually we could see "rounds{s,d} $0x0,%xmm0,%xmm0" where the mode is set to zero.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1597]:
```suggestion
	// Below, we use the same implementation: "rounds{s,d} $0x0,%xmm0,%xmm0" where the mode is set to zero.
```

File: wasm/jit/jit_amd64.go [hunk 14]
```diff
@@ -1440,6 +1456,192 @@ func (c *amd64Compiler) emitSimpleBinaryOp(instruction obj.As) error {
...
+	round.To.Reg = target.register
+	round.From.Type = obj.TYPE_CONST
+	round.From.Offset = mode
+	round.RestArgs = append(round.RestArgs,
+		obj.Addr{Reg: target.register, Type: obj.TYPE_REG})
+	c.addInstruction(round)
+	return nil
+}
+
+// compileSqrt emits the instructions to take the square root of the top value of float type on the stack.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64.go:1623]:
```suggestion
// compileSqrt adds instructions to replace the top value of float type on the stack with its square root.
```

File: wasm/jit/jit_amd64_test.go [hunk 15]
```diff
@@ -3877,6 +3877,215 @@ func TestAmd64Compiler_compileRem(t *testing.T) {
...
+	} {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			for i, v := range []uint64{
+				0,
+				1 << 63,
+				1<<63 | 12345,
+				1 << 31,
+				1<<31 | 123455,
+				6.8719476736e+10,
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:3908]: is this intentionally repeated?

File: wasm/jit/jit_amd64_test.go [hunk 16]
```diff
@@ -3877,6 +3877,215 @@ func TestAmd64Compiler_compileRem(t *testing.T) {
...
+								h := u / 2.0
+								if um < dm || (um == dm && float32(math.Floor(float64(h))) == h) {
+									f = u
+								} else {
+									f = d
+								}
+							}
+							expFloat32 = f
+						} else {
+							// Implementation from https://github.com/wasmerio/wasmer/blob/703bb4ee2ffb17b2929a194fc045a7e351b696e2/lib/vm/src/libcalls.rs#L120-L144.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_amd64_test.go:4021]: can you just re-implement these two? especially the variable names aren't great and I think we can do better

File: wasm/wazeroir/interpreter.go [hunk 17]
```diff
@@ -1212,6 +1212,7 @@ func (it *interpreter) callNativeFunc(f *interpreterFunction) {
 		case OperationKindNearest:
 			{
 				// Borrowed from https://github.com/wasmerio/wasmer/blob/703bb4ee2ffb17b2929a194fc045a7e351b696e2/lib/vm/src/libcalls.rs#L77
+				// TODO: use the new algorithm from https://github.com/bytecodealliance/wasmtime/pull/2171.
```
Comment (codefromthecrypt) [inline wasm/wazeroir/interpreter.go:1215]:
```suggestion
				// TODO: look at https://github.com/bytecodealliance/wasmtime/pull/2171 and reconsider this algorithm
```

File: wasm/wazeroir/interpreter.go (hunk 1 above)
Comment (codefromthecrypt) [reply wasm/wazeroir/interpreter.go:1226]: this is good, though this line is heavy and needs comment or better variable names.

**Review by codefromthecrypt — COMMENTED** [review]
Looks good.. some of the tests are feeling a little heavy though.. see if you can improve the readability perhaps by reworking them instead of copy.

File: wasm/jit/jit_amd64.go (hunk 12 above)
Comment (mathetake) [reply wasm/jit/jit_amd64.go]: https://github.com/tetratelabs/wazero/pull/119/commits/ee4aadab3e6f1faffe2cda117ec7e5ec17876bee added the example and also the test case for it!

File: wasm/jit/jit_amd64_test.go (hunk 16 above)
Comment (mathetake) [reply wasm/jit/jit_amd64_test.go]: re-implemented! https://github.com/tetratelabs/wazero/pull/119/commits/d6fc64a40917c4bb0b4277876a4c5b69afee680d

**Comment by mathetake** [comment]
@codefromthecrypt thanks for the suggestions!

**Comment by codefromthecrypt** [comment]
thanks for the good work!

**Pull request merged**